odoo/http.py

```diff
@@ -521,7 +521,7 @@ def response_wrap(*args, **kw):
                     _logger.info("<function %s.%s> called ignoring args %s" % (f.__module__, f.__name__, ', '.join(ignored)))
 
             response = f(*args, **kw)
-            if isinstance(response, Response) or f.routing_type == 'json':
+            if isinstance(response, Response) or f.routing_type in ['json', '*']:
```

Comment:
This means the case of an HTTP request won't be handled exactly the same with `type='*'` as with `type='http'`, e.g. next lines won't apply. Could we keep it functionally equivalent with something like this instead? 

```suggestion
            if isinstance(response, Response) or isinstance(request._get_current_object(), JsonRequest):
```

ref: https://werkzeug.palletsprojects.com/local/#werkzeug.local.LocalProxy._get_current_object

Reply:
@odony Hi :)
It doesn't always work, it seems that sometimes the "current object" is a python dict
See
https://runbot.odoo.com/runbot/build/1951024
![_tmp_screenshot](https://user-images.githubusercontent.com/54316226/77905386-7443b580-7286-11ea-8ddf-8fb27b8aa424.png)

Reply:
IMO it's safer to stay like this, to not impact existing routes

Reply:
The reason for this error is that `MockRequest` does not simulate a real request: https://github.com/odoo/odoo/blob/13.0/addons/website/tools.py#L96-L116
It's a bit of a pity to be held back in the core code because of implementation details of the test system.

What about introducing a `request.type` attribute then? It would be a very small change (example below), and would allow to test `request.type == 'json'` here instead of the suggested `isinstance()`, to the same effect.

And then you only have to adapt the few tests that call a json route with MockRequest, and pass `type='json'` explicitly.
It will also make a slightly faster and more elegant way of testing the request type than doing `hasattr(request, 'jsonrequest')` in other places.

The idea is something like this:
```diff
diff --git addons/website/tools.py addons/website/tools.py
index 3639776706fe..bbdb352d4859 100644
--- addons/website/tools.py
+++ addons/website/tools.py
@@ -112,6 +112,7 @@ class MockRequest(object):
                 'sale_order_id': kw.get('sale_order_id'),
             },
             'website': kw.get('website'),
+            'type': kw.get('type', 'http'),
         })
         odoo.http._request_stack.push(self.request)
 
diff --git odoo/http.py odoo/http.py
index 50874bfbd50f..337d8fbb3939 100644
--- odoo/http.py
+++ odoo/http.py
@@ -203,6 +203,7 @@ class WebRequest(object):
         self.endpoint = None
         self.endpoint_arguments = None
         self.auth_method = None
+        self.type = 'http'
         self._cr = None
         self._uid = None
         self._context = None
@@ -577,6 +578,7 @@ class JsonRequest(WebRequest):
     def __init__(self, *args):
         super(JsonRequest, self).__init__(*args)
 
+        self.type = 'json'
         self.params = {}
 
         args = self.httprequest.args

```

Reply:
@odony But so, that means that we can call only one "route type" in a `MockRequest` namespace.
Is it ok ?

Reply:
(and there's already the attribute `_request_type` :) no need to init it in the `__init__`)

Reply:
(and also, pay attention that the routing type is different than the request type, the request type can be "http" or "json" and the routing type can be "http", "json", or "*", it add a little confusion imo, but for a temporary solution it's ok)